<?xml version="1.0"?>
<launch>

  <param name="use_sim_time" value="false"/>
  <arg name="rviz" default="true" />
  <arg name="markerId" default="1"/>
  <arg name="meter_per_pixel" value="0.005"/>
  <arg name="image_width" value="2000" />
  <arg name="image_height" value="2000" />
  <!--arg name="pixel_mode" value="false"/-->
  <arg name="pixel_mode" value="false"/>

<!--   <node name="rqt_topic" pkg="rqt_topic" type="rqt_topic"/> -->
   <node name="rviz" pkg="rviz" type="rviz" args="-d $(find potential_field_navigation)/config/config_gazebo.rviz" output="screen" respawn="true"/>
  
    <!--Image-->
        <node name="image_selecter_node" pkg="potential_field_navigation" type="image_selecter_node" output="screen" cwd="ROS_HOME">
        <param name="image_publisher_topic" value="/image"/>
    </node>

    <!--Image postprocessing-->
        <node name="edge_separation_node" pkg="potential_field_navigation" type="edge_separation_node" output="screen" cwd="ROS_HOME">
        <param name="dyed_image_publisher_topic" value="/image/image_current"/>
        <param name="image_listener_topic" value="/image"/>
    </node>

    <!--Image postprocessing-->
        <node name="redyeing_charge_node" pkg="potential_field_navigation" type="redyeing_charge_node" output="screen" cwd="ROS_HOME">
        <param name="dyed_image_publisher_topic" value="/image/image_charge"/>
        <param name="image_listener_topic" value="/image"/>
    </node>

    <node name="image_to_vectorfield_current_node" pkg="potential_field_navigation" type="image_to_vectorfield_current_node" output="screen" cwd="ROS_HOME">
        <!--param name="image_listener_topic" value="/image"/-->
        <param name="image_listener_topic" value="/image/image_current"/>
        <!--param name="potentialfield_publisher_topic" value="/potentialfield/image"/-->
        <param name="vectorfield_publisher_topic" value="/vectorfield/final_image_current"/>
        <!--param name="vectorfield_publisher_topic" value="/vectorfield/final_image"/-->
        <param name="heuristic_abs_min" value="1.0"/> <?Undefined An arbitrary value which indecates the start of value decay?>
        <param name="heuristic_factor" value="80.0"/> <?Maximum vector lengh?>
        <!--param name="heuristic_apply" value="1"/> <?Use heuristic instead of native vectorfield?-->
        <param name="heuristic_apply" value="1"/> <?Use heuristic instead of native vectorfield?>
    </node>

    <node name="image_to_vectorfield_charge_node" pkg="potential_field_navigation" type="image_to_vectorfield_charge_node" output="screen" cwd="ROS_HOME">
      <!--param name="image_listener_topic" value="/image"/-->
      <param name="image_listener_topic" value="/image/image_charge"/>
      <param name="potentialfield_publisher_topic" value="/potentialfield/image"/>
      <param name="vectorfield_publisher_topic" value="/vectorfield/final_image_charge"/>
      <!--param name="vectorfield_publisher_topic" value="/vectorfield/final_image"/-->
      <param name="heuristic_abs_min" value="1.0"/> <?Undefined An arbitrary value which indecates the start of value decay?>
      <param name="heuristic_factor" value="70.0"/> <?Maximum vector lengh?>
      <!--param name="heuristic_apply" value="1"/> <?Use heuristic instead of native vectorfield?-->
      <param name="heuristic_apply" value="1"/> <?Use heuristic instead of native vectorfield?>
    </node>

    <!--Fusion-->
    <node name="vectorfield_merger_node" pkg="potential_field_navigation" type="vectorfield_merger_node" output="screen" cwd="ROS_HOME">
        <param name="field1_listener_topic" value="/vectorfield/final_image_current"/>
        <param name="field2_listener_topic" value="/vectorfield/final_image_charge"/>
        <param name="field_width" value="$(arg image_width)"/>
        <param name="field_height" value="$(arg image_height)"/>
        <param name="vectorfield_publisher_topic" value="/vectorfield/final_image"/>
    </node>

    <node name="vectorfield_to_gridmap_node" pkg="potential_field_navigation" type="vectorfield_to_gridmap_node" output="screen" cwd="ROS_HOME">
        <param name="meter_per_pixel" value="$(arg meter_per_pixel)"/>
        <param name="frame_id" value="world"/>
        <param name="vectorfield_listener_topic" value="/vectorfield/final_image"/>
        <param name="gridmap_publisher_topic" value="/vectorfield/image/gridmap"/>
    </node>

    <!--Steering-->
    <node name="amiro1_vectorfield_to_kinematic_node" pkg="potential_field_navigation" type="vectorfield_to_kinematic_node" output="screen" cwd="ROS_HOME">
        <param name="meter_per_pixel" value="$(arg meter_per_pixel)"/>
        <param name="vectorfield_listener_topic" value="/vectorfield/final_image"/>
        <param name="amiro_odom_listener_topic" value="/amiro1/odom"/>
        <param name="twist_publisher_topic" value="/amiro1/cmd_vel"/>
        <!--param name="velocityScale_meterPerSecond" value="0.1"/-->
        <param name="velocityScale_meterPerSecond" value="0.1"/>
        <!--param name="angularScale_radPerSecond" value="0.05"/-->
        <param name="angularScale_radPerSecond" value="1.0"/>
        <param name="pixel_mode" value="$(arg pixel_mode)"/>
        <param name="twist_mode" value="true"/>
    </node>

</launch>
